src/core/Internal.ml

```diff
@@ -546,81 +539,122 @@ module Make(Plugin : PLUGIN)
+            Clause.debug init (Format.pp_print_list Clause.debug) l);
+      Array.iter Atom.mark init.atoms;
+      let steps =
+        List.map
+          (fun c ->
+             let pivot =
+               match
+                 Sequence.of_array c.atoms
+                 |> Sequence.filter (fun a -> Atom.seen (Atom.neg a))
+                 |> Sequence.to_list
+               with
+                 | [a] -> a
+                 | [] ->
+                   error_res_f "(@[proof.expand.pivot_missing@ %a@])" Clause.debug c
+                 | pivots ->
+                   error_res_f "(@[proof.expand.multiple_pivots@ %a@ :pivots %a@])"
+                     Clause.debug c Atom.debug_l pivots
+             in
+             Array.iter Atom.mark c.atoms; (* add atoms to result *)
+             clear_var_of_ pivot;
+             Atom.abs pivot, c)
+          l
+      in
+      (* cleanup *)
+      let res = ref [] in
+      let cleanup_a_ a =
+        if Atom.seen a then (
+          res := a :: !res;
+          clear_var_of_ a
+        )
+      in
+      Array.iter cleanup_a_ init.atoms;
+      List.iter (fun c -> Array.iter cleanup_a_ c.atoms) l;
+      !res, steps
 
     let expand conclusion =
       Log.debugf 5 (fun k -> k "(@[sat.proof.expand@ @[%a@]@])" Clause.debug conclusion);
       match conclusion.cpremise with
       | Lemma l ->
-        {conclusion; step = Lemma l; }
+        { conclusion; step = Lemma l; }
       | Local ->
         { conclusion; step = Assumption; }
       | Hyp l ->
         { conclusion; step = Hypothesis l; }
       | History [] ->
         error_res_f "@[empty history for clause@ %a@]" Clause.debug conclusion
       | History [c] ->
-        let duplicates, res = analyze (list c) in
-        assert (cmp_cl res (list conclusion) = 0);
-        { conclusion; step = Duplicate (c, List.concat duplicates) }
-      | History ( c :: ([_] as r)) ->
-        let (l, c', d', a) = chain_res (c, to_list c) r in
-        assert (cmp_cl l (to_list conclusion) = 0);
-        { conclusion; step = Resolution (c', d', a); }
-      | History ( c :: r ) ->
-        let (l, c', d', a) = chain_res (c, to_list c) r in
-        conclusion.cpremise <- History [c'; d'];
-        assert (cmp_cl l (to_list conclusion) = 0);
-        { conclusion; step = Resolution (c', d', a); }
+        let duplicates, res = find_dups c in
+        assert (same_lits (Sequence.of_list res) (Clause.atoms_seq conclusion));
+        { conclusion; step = Duplicate (c, duplicates) }
+      | History (c :: ([_] as r)) ->
+        let res, steps = find_pivots c r in
+        assert (same_lits (Sequence.of_list res) (Clause.atoms_seq conclusion));
+        { conclusion; step = Hyper_res { hr_init=c; hr_steps=steps; }; }
+      | History (c :: r) ->
+        let res, steps = find_pivots c r in
+        assert (same_lits (Sequence.of_list res) (Clause.atoms_seq conclusion));
+        { conclusion; step = Hyper_res {hr_init=c; hr_steps=steps};  }
```

Comment:
These two last cases now do the exact same thing if I'm not mistaken, you can remove the duplication, ^^